docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,258 @@
+}
+
+// ImmutabilitySpec defines immutability settings.
+type ImmutabilitySpec struct {
+    // RetentionType indicates the type of immutability approach. For example, "Bucket".
+    RetentionType string `json:"retentionType,omitempty"`
+}
+```
+
+If `immutability` is not specified, `etcd-druid` will assume that the bucket is mutable, and no immutability-related logic applies.
+
+##### `etcd-backup-restore` Enhancements
+
+The authors propose adding two new commands to the `etcd-backup-restore` CLI (`etcdbrctl`) to maintain immutability during hibernation and to clean up older snapshots:
+
+1. **`reupload-snapshot`**
+   - Downloads the latest full snapshot from the object store.
+   - Renames the snapshot (for instance, updates its Unix timestamp) to avoid overwriting an existing immutable snapshot.
+   - Uploads the renamed snapshot back to object storage, thereby restarting its immutability timer.
+
+2. **`garbage-collect`**
+   - Scans the object store for older snapshots and deletes them if their immutability period has expired and they are no longer needed, following the standard [garbage collection policy](https://github.com/gardener/etcd-backup-restore/blob/master/docs/usage/garbage_collection.md).
+
+##### etcd Controller Enhancements
+
+When a hibernation flow is initiated (by external tooling or higher-level operators), the [etcd controller](https://github.com/gardener/etcd-druid/blob/master/docs/development/controllers.md#etcd-controller) can:
+
+1. Remove etcd’s client ports (2379/2380) from the etcd Service to block application traffic.
+2. Trigger an [on-demand full snapshot](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/05-etcd-operator-tasks.md#trigger-on-demand-fulldelta-snapshot) via an `EtcdOperatorTask`.
+3. Scale down the `StatefulSet` replicas to zero, provided the previous snapshot step is successful.
+4. Create the `ExtendFullSnapshotImmutabilityTask` if `etcd.spec.backup.store.immutability.retentionType` is `"Bucket"` and based on `etcd.spec.backup.fullSnapshotSchedule`.
+
+##### Operator Task Enhancements
+
+The `ExtendFullSnapshotImmutabilityTask` will create a cron job that:
+
+- Runs `etcdbrctl reupload-snapshot` to extend the immutability of the most recent snapshot.
+- Runs `etcdbrctl garbage-collect --garbage-collection-policy <policy>` to remove old snapshots.
+
+By periodically re-uploading the latest snapshot during hibernation, the authors ensure that the immutability period is extended, and the backups remain **protected throughout the hibernation period**.
+
+###### Example Task Config
+
+```go
+type ExtendFullSnapshotImmutabilityTaskConfig struct {
+  // Schedule defines a cron schedule (e.g., "0 */6 * * *").
+  Schedule *string `json:"schedule,omitempty"`
+
+  // GarbageCollectionConfig specifies the configuration for snapshot GC.
+  GarbageCollectionConfig *GarbageCollectionConfig `json:"garbageCollectionConfig,omitempty"`
+}
+
+type GarbageCollectionConfig struct {
```

Comment:
i did not quite understand this type in context of `ExtendFullSnapshotImmutabilityTaskConfig` - why is it that complicated? If GC needs to be done via this task as well then it should only be done for copied last full snapshot (taken prior to hibernation). I am confused why is delta snapshot retention coming here and what is exponential GC policy and how is relevant w.r.t this task.

Reply:
Now, I've removed the garbage collection config and ExtendFullSnapshotImmutabilityTask will automatically garbage collect the snapshots it has created, ensuring storage efficiency and preventing stale backups.